Concierge Chats: Add data layer handlers and actions for the fetch API of concierge shifts.

southp opened the pull request:
## Summary
This PR adds both the boilerplate code for the data layer handlers of the concierge APIs, and one specific handler for the API fetching concierge shifts.

## Test Plan
1. `npm run test-client client/state/data-layer/wpcom/concierge/` should pass.
1. `npm run test-client client/state/concierge` should pass.

dmsnell commented on client/state/action-types.js:
```diff
@@ -77,6 +77,9 @@ export const BILLING_TRANSACTIONS_RECEIVE = 'BILLING_TRANSACTIONS_RECEIVE';
 export const BILLING_TRANSACTIONS_REQUEST_SUCCESS = 'BILLING_TRANSACTIONS_REQUEST_SUCCESS';
+export const CONCIERGE_SHIFTS_FETCH = 'CONCIERGE_SHIFTS_FETCH';
+export const CONCIERGE_SHIFTS_FETCH_FAILED = 'CONCIERGE_SHIFTS_FETCH_FAILED';
+export const CONCIERGE_SHIFTS_FETCH_SUCCESS = 'CONCIERGE_SHIFTS_FETCH_SUCCESS';
```
it looks like this is a work in progress, but take note that we likely won't need these `_FAILURE` and `_SUCCESS` action types - `dispatchRequest()` reuses the original and in many cases our reducers can _infer_ the success/failure state from normal activity, or at a minimum we tend to dispatch something like a simple notice on failure and suggest that our customers retry their operation (without needing to track error state in the app)

mattwondra replied:
@dmsnell To make sure I understand — since the `http` helper adds succes/error metadata to the action before passing it back to the handlers, we don't need the `_SUCCESS`/`_FAILURE` actions because we can simply re-dispatch `CONCIERGE_SHIFTS_FETCH` but with `meta.dataLayer.data` or `meta.dataLayer.error` auto-attached?

So we'd expect the reducers to see `CONCIERGE_SHIFTS_FETCH` twice, once for the initial fetch and once for the success or error, but they can differentiate based on the metadata in the action.

And assuming we don't need any special logic, we can `bypassDataLayer` on the action we pass to the `http` helper so that the success/error dispatch doesn't even come back through the data layer, right?

```js
import { bypassDataLayer } from 'state/data-layer/utils';
import { http } from 'state/data-layer/wpcom-http/actions';
import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
import { CONCIERGE_SHIFTS_FETCH } from 'state/action-types';

export const handleFetch = ( { dispatch }, action ) => {
	// ...
	dispatch(
		http(
			{ /* request params ... */ },
			// Bypass the data layer on the action we pass in, so that when
			// the success/error metadata is attached and it's re-dispatched
			// it won't be processed here at all
			bypassDataLayer( action )
		)
	);
};

export default {
	[ CONCIERGE_SHIFTS_FETCH ]: [
		// We don't need to pass success/error handlers to dispatchRequest
		// because the success/error action will bypass the data layer
		dispatchRequest( handleFetch )
	],
};
```

I'm having a hard time grokking everything since I haven't worked with the data layer in a while and it looks like there's been changes in the past week that make the docs outdated. Please let me know if this is the idea you were thinking about, or if I'm way off :)

dmsnell replied:
Hopefully it's simpler than it seems here. Ideally we have separate semantic action types for _requesting_ data from the server and _actually adding in data to the store_. Think of it as the difference between asking for a list of resumes and posting resumes to the hiring queue.

In this case we might expect a kind of `CONCIERGE_SHIFTS_REQUEST` to indicate to Calypso that we need the list of shifts. The reducers will likely ignore this action. It's mostly meaningless to the app itself. What matters instead to the app is stuff like, `CONCIERGE_SHIFT_ADD` or `CONCIERGE_SHIFT_REMOVE` or `CONCIERGE_SHIFT_UPDATE` - things that we might be doing regardless of any network activity. For example, we might run this from the console dispatcher in order to test a specific app state while working on the UI.

So the data layer with `dispatchRequest()` can reuse the `_REQUEST` action for the lifecycle of the network request. It may not go out immediately or at all (which is what differentiates a _request_ from a _fetch_). When it comes back in we instruct the data layer how to translate that API data into meaningful actions for Calypso.

Thus we _will_ have three functions at play here: a function to initiate network activity (by describing a network request), a function to generate Redux actions when the network activity is successful, and a function to generate actions when it fails. The data layer will do its own work on the network bits to try and keep things moving smoothly across temporary network failures and it will prevent sending out the same request at the same time.

```js
dispatchRequest( fetchShifts, storeShifts, announceFailure )
```

^^^ that's a typical pattern. `fetchShifts()` dispatches an `http()` action (which needs no bypass - the _request_ action itself should be handled here on return), the `storeShifts()` translates data from the API into data that makes sense for Calypso and dispatches those actions, then the announcer typically just sets up an `errorNotice()` indicating things failed. It's nice to alert our customers of the failure, and in most cases it's easiest just to try again so that we don't try to track complicated failure state.

The [countries list](https://github.com/Automattic/wp-calypso/blob/ff03397bbca59c5d56cfbc494123d9ddc7fa7a7d/client/state/data-layer/wpcom/domains/countries-list/index.js) is a simple example.

Thanks for poking me to ask about this. I'll be happy to answer any questions you may have.

mattwondra replied:
@dmsnell 👍 Awesome, that's a very clear explanation. Thanks for taking the time.

@southp In light of that, this PR could probably use some updates, so that we only have the one action representing the network request, and then probably just one more that takes the fetched data and passes it to reducers. (I'm not sure from a UI perspective how we want to handle errors yet, so we may just want to stub out an error handler that doesn't do anything at this point).

mattwondra commented on client/state/data-layer/wpcom/concierge/shifts/index.js:
```diff
@@ -0,0 +1,36 @@
+	[ CONCIERGE_SHIFTS_FETCH ]: [
+		dispatchRequest(
+			handleFetchConciergeShifts,
+			fetchConciergeShiftsSuccess,
```
See comments above with @dmsnell (we may want to ditch special success/error actions) but if we do want to keep them, you'll need to specifically dispatch them. This newer `dispatchRequest` helper doesn't auto-dispatch anything, so you can't just pass in an action, you need to create a special handler that actually dispatches the action.

southp commented:
@dmsnell @mattwondra Thanks for the insightful feedback : )
I have updated the code to fit in the pattern suggested in the above comments. Mainly:

* Replace fetch / fetch success / fetch failed actions by simply a fetch action for the request and a update action for storing the data.
* Replace the error handler by the one dispatching error notices.

Please help me have another look to see if I misunderstand anything 🙇🏻

dmsnell commented on client/state/data-layer/wpcom/concierge/shifts/index.js:
```diff
@@ -0,0 +1,54 @@
+	);
+
+export default {
+	[ CONCIERGE_SHIFTS_FETCH ]: [
```
Just a note here somewhat mentioned before, but our nomenclature for the data layer code stresses `_REQUEST` action suffixes over `_FETCH` to distinguish between actions which actually force something to happen and those which indicate a data need which Calypso will try to fulfill.

In other words, there's a tendency to see a 1:1 correlation between `_FETCH` action dispatches and network requests, but no such relationship between `_REQUEST` action dispatches and network requests. `_REQUEST` is saying, "I need this data," while `_FETCH` communicates, "Get this data now"

southp replied:
Ah, right. That's something that I missed. I've updated the naming in b7ce14b and 1675197.

dmsnell commented on client/state/data-layer/wpcom/concierge/shifts/index.js:
```diff
@@ -0,0 +1,54 @@
+				method: 'GET',
+				path: `/concierge/schedules/${ scheduleId }/shifts`,
+				apiNamespace: 'wpcom/v2',
+				retryPolicy: noRetry(),
```
Any reason for not retrying? If so, it should probably be well-documented here.

southp replied:
I didn't find uses of the `exponentialBackoff` policy, so I thought it was a work-in-progress. It is using `exponentialBackoff` now after b71a08a.

dmsnell replied:
Actually the exponential backoff policy is the default - there's no need to provide any `retryPolicy` unless we want to intentionally change it from the default.

southp replied:
Somehow I missed it as the default. It is now removed by 8c403b0.

dmsnell commented:
@southp since the last time we worked together in the data layer things have changed a bit (hopefully not too much). This code is looking good but I wanted to let you know of some developments which you helped pioneer! There's now an official `fromApi` optional argument to `dispatchRequest` which can be used to validate and transform API responses.

While not required, if you use it then Calypso can validate an API response and/or transform its shape for you with as little as a provided JSONSchema.

```
const responseSchema = {
  type: 'object',
  properties: {
    shifts: {
      type: 'array',
      items: { … }
    }
  }
}

[ CONCIERGE_SHIFTS_REQUEST ]: [ dispatchRequest(
  fetch,
  store,
  announce,
  { fromApi: makeParser( responseSchema ) }
)
```

or something like that. have fun!

southp commented:
@dmsnell Thanks for guiding me through the state-of-the-art data layer practices! I've updated the code for addressing your comments above. I'd be grateful if you can come back and have the 3rd round of review :)

About the `fromApi` and the response validating schema, while I'm so eager to add them right in this PR, I think it would be better to leave that in the next PR to avoid further bloat the size of this :)

dmsnell commented on client/state/data-layer/wpcom/concierge/shifts/test/index.js:
```diff
@@ -0,0 +1,76 @@
+						status: 'is-error',
+					} ),
+				} )
+			);
```
notice that if we want to verify the the error is created we may consider doing so in a way that doesn't hard-code the error action itself. instead, we can call `errorNotice()` here in the test itself. there's a unique id it generates but we can additionally pass that in ourselves or we can mock `impure-lodash#uniqueId` to force a test value (something >1)

southp replied:
Nice catch :)
The other problem of that code was that it introduced more dependencies than the code it tested, which is a wrong design in most of the cases in my opinion. i.e. `/concierge/shifts.js` has zero knowledge of `NOTICE_CREATE` action type, but the test used it to verify its behavior. 
I've updated it in 71e445f. Since I personally don't want to repeat the same error message in both the module and the test suite, I extracted the error notice action as a exported object.

How do you think? :)

dmsnell replied:
yeah it's fine. some of the deeper question is probably how valuable the test is itself. that's not easily answerable I don't think. you did a great job either way!

dmsnell replied:
also in case you are curious, [here](https://github.com/Automattic/wp-calypso/blob/8c75edbbc9d95b2e633af4707abf7706815c75e3/client/lib/media/test/utils.js#L18) is an example of how we can mock out `impure-lodash` to force a specific message id

on the other side, we can also [force the id](https://github.com/Automattic/wp-calypso/blob/8c75edbbc9d95b2e633af4707abf7706815c75e3/client/state/data-layer/wpcom/sites/activity/test/index.js#L85)

southp replied:
Ah, I didn't get it when you first mention that we can mock out `impure-lodash`. Thanks for finding the snippet for me.
I've applied it in 70d3051. Does it look good yo you? :)

dmsnell commented on client/state/data-layer/wpcom/concierge/shifts/index.js:
```diff
@@ -0,0 +1,50 @@
+/** @format */
```
this may seem quite obtuse, but normally we would expect to find this file at `…/wpcom/concierge/schedules/shifts` because of the actual API path it calls. The idea is that we can work background from network call to calling code.

southp replied:
That's reasonable. I've moved them in c12feb2.

dmsnell commented on client/state/data-layer/wpcom/concierge/shifts/index.js:
```diff
@@ -0,0 +1,50 @@
+
+export const conciergeShiftsFetchErrorNotice = errorNotice(
+	translate( "We've encountered problems trying to load available shifts. Please try again later." )
+);
```
Sadly this is also going against the grain of how translations work. Honestly I think it would probably be fine, but what it would end up in fact is a string that is only translated once at boot time and never updated. We have options around it, the simplest being making it a function…

```js
export const fetchError = () => errorNotice( translate( … ) );
```

now when we call it we run it and the translation will be produced on the spot in whatever the current language is.

southp replied:
OOps. That's quite vital I believe. In the end, I decided to go to the old way with a clearer explanation in 6168064, since I think it's a better compromise to work around the auto-incremented underlying `noticeId`. Now I can finally understand why [countries list](https://github.com/Automattic/wp-calypso/blob/ff03397bbca59c5d56cfbc494123d9ddc7fa7a7d/client/state/data-layer/wpcom/domains/countries-list/test/index.js#L65) is doing this ...

dmsnell commented:
This is nice work @southp - sorry to keep leaving little notes on it.

southp commented:
@dmsnell I always appreciate your craftsmanship (codermanship?) on thinking through every detail, so no worries :)
I've addressed your latest comments, so it would be great if you can revisit :)

mattwondra commented on client/state/data-layer/wpcom/concierge/schedules/shifts/index.js:
```diff
@@ -0,0 +1,52 @@
+	dispatch(
+		errorNotice(
+			translate(
+				"We've encountered problems trying to load available shifts. Please try again later."
```
I don't think this is the best error message to show customers, who are landing into the Concierge scheduler with no context about what a "shift" is. Maybe instead:

> We couldn't load our Concierge schedule. Please try again later.

southp replied:
Ah, that's better :)
Updated in 4e9f5d1.

dmsnell commented on client/state/data-layer/wpcom/concierge/schedules/shifts/test/index.js:
```diff
@@ -0,0 +1,65 @@
+
+// we are mocking impure-lodash here, so that conciergeShiftsFetchError() will contain the expected id in the tests
+jest.mock( 'lib/impure-lodash', () => ( {
+	uniqueId: () => 'mock-unique-id',
```
nice job not having it return a `0` or a `1` - this is important because the default behavior will return a `1` and we can sometimes get false positives in our tests because we accidentally hid a bug in our mocks 🙃
